libraries/libfc/src/string.cpp

```diff
@@ -83,13 +85,42 @@ namespace fc  {
   std::string trim( const std::string& s )
   {
       return boost::algorithm::trim_copy(s);
-      /*
-      std::string cpy(s);
-      boost::algorithm::trim(cpy);
-      return cpy;
-      */
   }
 
+  std::pair<std::string&, bool> escape_str( std::string& str, escape_control_chars escape_ctrl,
+                                            std::size_t max_len, std::string_view add_truncate_str )
+  {
+     bool modified = false, truncated = false;
+     // truncate early to speed up escape
+     if (str.size() > max_len) {
+        str.resize(max_len);
+        modified = truncated = true;
+     }
+     auto itr = escape_ctrl == escape_control_chars::on
+           ? std::find_if(str.begin(), str.end(),
+                          [](const auto& c) {
+              return c == '\x7f' || c == '\\' || c == '\"' ||  (c >= '\x00' && c <= '\x1f'); } )
+           : std::find_if(str.begin(), str.end(),
+                          [](const auto& c) {               // x09 = \t, x0a = \n,                   x0d = \r
+              return c == '\x7f' || (c >= '\x00' && c <= '\x08') || c == '\x0b' || c == '\x0c' || (c >= '\x0e' && c <= '\x1f'); } );
+
+     if (itr != str.end() || !fc::is_valid_utf8( str )) {
+        str = escape_string(str, nullptr, escape_ctrl == escape_control_chars::on);
+        modified = true;
+        if (str.size() > max_len) {
+           str.resize(max_len);
+           truncated = true;
+        }
+     }
+
+     if (truncated && !add_truncate_str.empty()) {
```

Comment:
```suggestion
     if (truncated) {
```

Reply:
I thought about that, but the impl of `string.append()` would make additional `if` checks not needed.

Reply:
oh, but will almost never be empty(), I'll remove the check.

Reply:
[operator+=](https://en.cppreference.com/w/cpp/string/basic_string/operator%2B%3D) has a version that takes a StringViewLike, so the test would de done there instread of in the `if` before. I'm not seeing any drawback.

Reply:
The impl of the `operator+=` that takes a `string_view` does:
`append(sv.data(), sv.size())` and then it checks `if (size)` in the impl of `append`. Regardless, I agree and have removed the check.